app/test/testformeditors.cpp

```diff
@@ -156,3 +157,67 @@ void TestFormEditors::testNumericFields()
   // field "Cabin Crew" stayed with invalid input, check controller flag of values validity
   QCOMPARE( controller.fieldValuesValid(), false );
 }
+
+void TestFormEditors::testRelationsWidgetPresence()
+{
+  QString projectDir = TestUtils::testDataDir() + "/planes";
+  QString projectName = "quickapp_project.qgs";
+
+  QString layersDb = projectDir + "/geo-layers.gpkg";
+
+  QSignalSpy spy( QgsProject::instance()->relationManager(), &QgsRelationManager::relationsLoaded );
+  QVERIFY( QgsProject::instance()->read( projectDir + "/" + projectName ) );
+  QCOMPARE( spy.count(), 1 );
+
+
+  QgsMapLayer *airportsLayer = QgsProject::instance()->mapLayersByName( QStringLiteral( "airports" ) ).at( 0 );
+  QgsVectorLayer *airportsVLayer = static_cast<QgsVectorLayer *>( airportsLayer );
+
+  QgsMapLayer *airportTowersLayer = QgsProject::instance()->mapLayersByName( QStringLiteral( "airport-towers" ) ).at( 0 );
+  QgsVectorLayer *airportTowersVLayer = static_cast<QgsVectorLayer *>( airportTowersLayer );
+
+  QVERIFY( airportsVLayer && airportsVLayer->isValid() );
+  QVERIFY( airportTowersVLayer && airportTowersVLayer->isValid() );
+
+  bool containsRelation = false;
+  bool containsRelationReference = false;
+
+  // check if we added relation widget when project has autogenerated layout
+
+  QgsFeature f = airportsVLayer->getFeature( 1 );
+  FeatureLayerPair pair( f, airportsVLayer );
+
+  AttributeController controller;
+  controller.setFeatureLayerPair( pair );
+
+  const TabItem *tabItem = controller.tabItem( 0 );
+  QVector<QUuid> formItems = tabItem->formItems();
+
+  for ( const auto &itemId : formItems )
+  {
+    const FormItem *item = controller.formItem( itemId );
+    if ( item->editorWidgetType() == QStringLiteral( "relation" ) )
+      containsRelation = true;
+  }
```

Comment:
would be good to check there's exactly one relation + there's no relation ref widget
(and the other way around for the other layer)